zulip_bots/zulip_bots/lib.py

```diff
@@ -115,13 +114,73 @@ def get(self, key: Text) -> Any:
         self.state_[key] = marshalled_value
         return self.demarshal(marshalled_value)
 
+    @contextmanager
+    def use(self, keys: List[Text]) -> Iterator[BaseStorage]:
+        # The context manager for StateHandler that minimizes the number of round-trips to the server.
+        # It will fetch all the data using the specified keys and store them to
+        # a CachedStorage that will not communicate with the server until manually
+        # calling flush or getting some values that are not previously fetched.
+        not_cached = list(filter(lambda key: key not in self.state_.keys(), keys))
+        response = self._client.get_storage({'keys': not_cached})
+        if response['result'] != 'success':
+            raise KeyError('One or more of the keys are not found: ' + ', '.join(not_cached))
+        data = {key: self.demarshal(value) for key, value in response['storage'].items()}
+
+        cache = CachedStorage(self, data)
+        yield cache
+        cache.flush()
+
     def contains(self, key: Text) -> bool:
         return key in self.state_
 
-    @contextmanager
+class CachedStorage:
+    def __init__(self, parent_storage: BotStorage, init_data: Dict[str, Any]) -> None:
+        # CachedStorage is implemented solely for the context manager of any BotHandler.
+        # It has a parent_storage that is responsible of communicating with the database
+        #   1. when certain data is not cached;
+        #   2. when the data need to be flushed to the database.
+        # It can be initialized with the given data.
+        self._parent_storage = parent_storage
+        self._cache = init_data
+        self._dirty_keys: Set[str] = set()
+
+    def put(self, key: Text, value: Any) -> None:
+        # In the cached storage, values being put to the storage is not flushed to the parent storage.
+        # It will be marked dirty until it get flushed.
+        self._cache[key] = value
+        self._dirty_keys.add(key)
+
+    def get(self, key: Text) -> Any:
+        # Unless the key is not found in the cache, the cached storage will not lookup the parent storage.
+        if key in self._cache:
+            return self._cache[key]
+        else:
+            value = self._parent_storage.get(key)
+            self._cache[key] = value
+            return value
+
+    def flush(self) -> None:
+        # Flush the data to the parent storage.
+        # Data that are not marked dirty will be omitted.
+        # This should be manually called when CachedStorage is not used with a context manager.
+        while len(self._dirty_keys) > 0:
+            key = self._dirty_keys.pop()
+            self._parent_storage.put(key, self._cache[key])
+
+    def flush_one(self, key: Text) -> None:
+        # Flush data with the specified key to the server.
+        self._dirty_keys.remove(key)
+        self._parent_storage.put(key, self._cache[key])
+
     def use(self, keys: List[Text]) -> Iterator[BaseStorage]:
```

Comment:
I am not sure why do we need this since the `CachedStorage` does not follow any `Protocol`. And am not sure if we explicitly need a `Protocol` for this.

Reply:
That makes sense. perhaps I can simply delete it.

Reply:
BTW, I later found out that this is required as the `CachedStorage` will otherwise not satisfy the `BotStorage` protocol.